mountpoint-s3/src/prefetch.rs

```
@@ -289,80 +274,53 @@ where
+    fn get_next_request_size(&self, request_size: usize) -> usize {
         // TODO: this logic doesn't work well right now in the case where part_size <
         // first_request_size and sequential_prefetch_multiplier = 1. It ends up just repeatedly
         // shrinking the request size until it reaches 1. But this isn't a configuration we
         // currently expect to ever run in (part_size will always be >= 5MB for MPU reasons, and a
         // prefetcher with multiplier 1 is not very good).
-        let next_request_size = (self.next_request_size * self.inner.config.sequential_prefetch_multiplier)
-            .min(self.inner.config.max_request_size);
-        self.inner
-            .part_feed
-            .get_aligned_request_size(self.next_request_offset, next_request_size)
+        (request_size * self.inner.config.sequential_prefetch_multiplier).min(self.inner.config.max_request_size)
     }
 
     /// Reset this prefetch request to a new offset, clearing any existing tasks queued.
     fn reset_prefetch_to_offset(&mut self, offset: u64) {
         self.current_task = None;
         self.future_tasks.drain(..);
         self.backward_seek_window.clear();
-        self.next_request_size = self.inner.config.first_request_size;
         self.next_sequential_read_offset = offset;
+        self.next_request_size = self.inner.config.first_request_size;
         self.next_request_offset = offset;
     }
 
     /// Try to seek within the current inflight requests without restarting them. Returns true if
     /// the seek succeeded, in which case self.next_sequential_read_offset will be updated to the
     /// new offset. If this returns false, the prefetcher is in an unknown state and must be reset.
-    async fn try_seek(&mut self, offset: u64) -> Result<bool, PrefetchReadError<TaskError<Client>>> {
+    async fn try_seek(&mut self, offset: u64) -> Result<bool, ObjectClientError<PrefetchReadError, Feed::ClientError>> {
```

This error type seems inverted — seeking should return a prefetcher-level error, one of the causes of which (the only cause of which?) is a client error.

---

If I understand correctly, you mean `try_seek` (and the `read` method) should return an error about like this:

```rust
enum PrefetchReadError<ClientError> {
  GetRequestFailed(ObjectClientError<GetObjectError,ClientError>),
  ..
  Integrity(IntegrityError),
}
```
which is not too far from what we had before, where `type TaskError<Client> = ObjectClientError<GetObjectError, <Client as ObjectClient>::ClientError>`.

My thinking, though, is that using `ObjectClientError` here with `PrefetchReadError` as the "`ServiceError`" fits nicely with the other `ObjectStore`/`ObjectClient` methods. Compare for example:

```rust
ObjectStore::head_object(..) -> ObjectClientResult<HeadObjectResult, HeadObjectError, Self::ClientError>
PrefetchGetObject::read(..) -> ObjectClientResult<ChecksummedBytes, PrefetchReadError, Self::ClientError>
```
It also means we don't need `PrefetchReadError` to be generic at all, but just wrap a `GetObjectError`.